### PR TITLE
feat: add parent asset id and initial loading status to asset tree

### PR DESCRIPTION
**Pull request by ghost**
The use case for this is when you want to show the hierarchy for a specific asset(s). The hierarchy tree only shows starting from root, which may not be the behavior thats needed. This is useful for Discovery, which will add this right after :)

### CODE REVIEW

**Comment by cognite-cicd** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-490.gearboxjs.preview.cogniteapp.com

**Comment by cognite-cicd** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-490.gearboxjs.preview.cogniteapp.com

File: src/components/AssetTree/AssetTree.tsx [hunk 1]
```diff
@@ -173,18 +195,24 @@ class AssetTree extends React.Component<AssetTreeProps, AssetTreeState> {
...
-      <Tree
-        loadData={onLoadData}
-        onSelect={(_, e) =>
-          onSelectNode(e.node.props.title as string, e.node.props.eventKey)
-        }
-        expandedKeys={Object.keys(expandedKeys)}
-        onExpand={onExpand}
-        children={renderTreeNode(rootAssetNodes)}
-      />
+      <>
```
Comment (polomani) [inline src/components/AssetTree/AssetTree.tsx]: why did you wrap it in here?

**Review by polomani — COMMENTED** [review]
Looks good, but could you explain a usecase for this one?
I am thinking maybe we should have something like `assetIds` property instead, and render those assets and everything under them. in that case it's more flexible (like you don't need to have parent for them, so you can render some filtered root nodes)

File: src/components/AssetTree/AssetTree.tsx (hunk 1 above)
Comment (dlcognite) [reply src/components/AssetTree/AssetTree.tsx]: Ahh i was testing thanks for the catch!

**Comment by dlcognite** [comment]
Fixed based on comments :) thanks for catching all of these!

**Comment by cognite-cicd** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-490.gearboxjs.preview.cogniteapp.com

**Comment by cognite-cicd** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-490.gearboxjs.preview.cogniteapp.com

File: src/components/AssetTree/stories/assetIds.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import 'antd/dist/antd.css';
+
+import React from 'react';
+import { AssetTree } from '@cognite/gearbox';
+
+function ExampleComponent(props) {
+
+  return (
+    <AssetTree
+      assetIds={1234}
```
Comment (mrtraser) [inline src/components/AssetTree/stories/assetIds.md]: Here we should pass an array of ids I guess

File: src/components/AssetTree/stories/full.md [hunk 3]
```diff
@@ -34,7 +34,9 @@ No required props
 
 | Property              | Description                                 | Type                        | Default |
 | --------------------- | ------------------------------------------- | --------------------------- | ------- |
-| `displayName`         | Modify the display name of an asset         | `(asset: Asset) => string` | `(...) => 'assetName: assetDescription' |
+| `displayName`         | Modify the display name of an asset         | `(asset: Asset) => string` | `(...) => 'assetName: assetDescription'` |
+| `showLoading`         | Show loading animation when fetching original list         | `boolean ` | `false` |
+| `parentAssetId`       | Show only assets under the specified parent id         | `number (InternalID)` |  |
```
Comment (mrtraser) [inline src/components/AssetTree/stories/full.md]: Prop name should be fixed as well as a type

File: src/components/AssetTree/AssetTree.tsx [hunk 4]
```diff
@@ -76,29 +77,47 @@ class AssetTree extends React.Component<AssetTreeProps, AssetTreeState> {
...
     };
   }
 
   async componentDidMount() {
+    this.loadAssetInfo();
+  }
+
+  componentDidUpdate(prevProps: AssetTreeProps) {
+    if (
+      JSON.stringify(prevProps.assetIds) !== JSON.stringify(this.props.assetIds)
```
Comment (mrtraser) [inline src/components/AssetTree/AssetTree.tsx]: Mb we can use `_.isEqual()` here?

**Comment by cognite-cicd** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-490.gearboxjs.preview.cogniteapp.com

**Comment by cognite-cicd** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-490.gearboxjs.preview.cogniteapp.com

File: src/components/AssetTree/AssetTree.tsx [hunk 5]
```diff
@@ -76,29 +78,45 @@ class AssetTree extends React.Component<AssetTreeProps, AssetTreeState> {
...
-    const assetNodesMap = AssetTree.convertAssetsToNodeMap(assets);
-    const rootAssetNodes = Object.keys(assetNodesMap).map(Number);
-
-    this.setState({
-      rootAssetNodes,
-      assetNodesMap,
+    this.setState({ loading: true }, async () => {
+      const { assetIds } = this.props;
+      const assets = await this.context!.assets.list({
+        filter: assetIds ? { parentIds: assetIds } : { root: true },
```
Comment (polomani) [inline src/components/AssetTree/AssetTree.tsx]: `assetIds ? { parentIds: assetIds }` this is wrong? you have to call https://docs.cognite.com/api/v1/#operation/byIdsAssets in this case instead

File: src/components/AssetTree/stories/assetIds.md [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+import 'antd/dist/antd.css';
+
+import React from 'react';
+import { AssetTree } from '@cognite/gearbox';
+
+function ExampleComponent(props) {
+
+  return (
+    <AssetTree
+      assetIds=[1234]
```
Comment (polomani) [inline src/components/AssetTree/stories/assetIds.md]:
```suggestion
      assetIds={[1234]}
```

File: src/components/AssetTree/AssetTree.tsx (hunk 5 above)
Comment (dlcognite) [reply src/components/AssetTree/AssetTree.tsx]: how is this wrong? you only want to show the children of the list of assets you've provided

Comment (dlcognite) [reply src/components/AssetTree/AssetTree.tsx]: For discovery, for instance, I want to show all assets under IAA in this format, i would pass it the id of IAA.

**Comment by cognite-cicd** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-490.gearboxjs.preview.cogniteapp.com

**Comment by cognite-cicd** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-490.gearboxjs.preview.cogniteapp.com

**Comment by cognite-cicd** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-490.gearboxjs.preview.cogniteapp.com

File: src/components/AssetTree/AssetTree.tsx (hunk 5 above)
Comment (dlcognite) [reply src/components/AssetTree/AssetTree.tsx]: Decided on a new logic!

File: src/components/AssetTree/AssetTree.tsx [hunk 7]
```diff
@@ -76,29 +78,56 @@ class AssetTree extends React.Component<AssetTreeProps, AssetTreeState> {
...
-    const assets = await this.context.assets
-      .list({ filter: { root: true } })
-      .autoPagingToArray(this.autoPagingToArrayOptions);
+    this.setState({ loading: true }, async () => {
+      const { assetIds } = this.props;
+      let assets: Asset[] = [];
+      if (assetIds) {
+        if (assetIds.length > 0) {
+          assets = await this.context!.assets.retrieve(
+            assetIds.map(el => ({ id: el }))
```
Comment (polomani) [inline src/components/AssetTree/AssetTree.tsx]:
```suggestion
            assetIds.map(id => ({ id }))
```

File: src/components/AssetTree/AssetTree.tsx [hunk 8]
```diff
@@ -76,29 +78,56 @@ class AssetTree extends React.Component<AssetTreeProps, AssetTreeState> {
...
+          );
+        }
+      } else {
+        assets = await this.context!.assets.list({
+          filter: assetIds ? { parentIds: assetIds } : { root: true },
```
Comment (polomani) [inline src/components/AssetTree/AssetTree.tsx]:
```suggestion
          filter: { root: true },
```

File: src/components/AssetTree/AssetTree.tsx [hunk 9]
```diff
@@ -76,29 +78,56 @@ class AssetTree extends React.Component<AssetTreeProps, AssetTreeState> {
...
+          aggregatedProperties: ['childCount'],
+        }).autoPagingToArray(this.autoPagingToArrayOptions);
+      }
 
-    const assetNodesMap = AssetTree.convertAssetsToNodeMap(assets);
-    const rootAssetNodes = Object.keys(assetNodesMap).map(Number);
+      console.log(assets);
```
Comment (polomani) [inline src/components/AssetTree/AssetTree.tsx]: remove console.log, please

File: src/components/AssetTree/stories/full.md [hunk 10]
```diff
@@ -34,7 +34,9 @@ No required props
 
 | Property              | Description                                 | Type                        | Default |
 | --------------------- | ------------------------------------------- | --------------------------- | ------- |
-| `displayName`         | Modify the display name of an asset         | `(asset: Asset) => string` | `(...) => 'assetName: assetDescription' |
+| `displayName`         | Modify the display name of an asset         | `(asset: Asset) => string` | `(...) => 'assetName: assetDescription'` |
+| `showLoading`         | Show loading animation when fetching original list         | `boolean ` | `false` |
+| `assetIds`       | Show only assets that are children of the specified ids (max 1000)         | `number (InternalID)[]` |  |
```
Comment (polomani) [inline src/components/AssetTree/stories/full.md]:
```suggestion
| `assetIds`       | To display a tree with provided assets as root nodes         | `number (InternalID)[]` |  |
```

File: src/components/AssetTree/stories/assetIds.md [hunk 11]
```diff
@@ -0,0 +1,22 @@
+## Parent Asset ID
```
Comment (polomani) [inline src/components/AssetTree/stories/assetIds.md]:
```suggestion
## With asset IDs
```

File: src/components/AssetTree/stories/AssetTree.stories.tsx [hunk 12]
```diff
@@ -127,6 +128,17 @@ storiesOf('AssetTree/Examples', module)
       },
     }
   )
+  .add(
+    'Asset IDs',
```
Comment (polomani) [inline src/components/AssetTree/stories/AssetTree.stories.tsx]:
```suggestion
    'With asset IDs',
```

**Comment by polomani** [comment]
it doesn't render in a storybook btw

**Comment by cognite-cicd** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-490.gearboxjs.preview.cogniteapp.com

File: package.json [hunk 13]
```diff
@@ -50,7 +50,7 @@
     "@babel/preset-env": "^7.3.1",
     "@babel/preset-react": "^7.0.0",
     "@cognite/griff-react": "~0.4.2",
-    "@cognite/sdk": "^2.7.0",
```
Comment (polomani) [inline package.json]: what's the reason for this?

**Comment by dlcognite** [comment]
Opps local changes getting pushed

**Comment by cognite-cicd** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-490.gearboxjs.preview.cogniteapp.com

**Comment by cognite-cicd** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-490.gearboxjs.preview.cogniteapp.com

**Comment by cognite-cicd** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-490.gearboxjs.preview.cogniteapp.com

**Review by polomani — APPROVED** [review]
LGTM now :)

**Review by mrtraser — APPROVED** [review]

**Comment by polomani** [comment]
@dlcognite remember to squash

**Comment by cognite-cicd** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-490.gearboxjs.preview.cogniteapp.com

**Pull request merged**